File: 40_implement-queue-by-two-stacks/implement-queue-by-two-stacks.py
# coding:utf-8
'''
@Copyright:LintCode
@Author:   ultimate010
@Problem:  http://www.lintcode.com/problem/implement-queue-by-two-stacks
@Language: Python
@Datetime: 16-06-14 08:16
'''

class MyQueue:

    def __init__(self):
        self.stack1 = []
        self.stack2 = []
        
    def push(self, element):
        # write your code here
        self.stack1.append(element)

    def top(self):
        # write your code here
        # return the top element
        if len(self.stack2) > 0:
            return self.stack2[-1]
        
        while len(self.stack1) > 0:
            self.stack2.append(self.stack1.pop())
            
        return self.stack2[-1]

    def pop(self):
        # write your code here
        # pop and return the top element
        if len(self.stack2) > 0:
            return self.stack2.pop()
        while len(self.stack1) > 1:
            self.stack2.append(self.stack1.pop())
        ret = self.stack1.pop()
        return ret

File: 40_implement-queue-by-two-stacks/implement-queue-by-two-stacks.cpp
/*
@Copyright:LintCode
@Author:   ultimate010
@Problem:  http://www.lintcode.com/problem/implement-queue-by-two-stacks
@Language: C++
@Datetime: 16-08-05 13:35
*/

class Queue {
public:
    stack<int> stack1;
    stack<int> stack2;

    Queue() {
        // do intialization if necessary
    }

    void push(int element) {
        // write your code here
        stack1.push(element);
    }
    
    int pop() {
        // write your code here
        if (stack2.size() > 0) {
            int ret = stack2.top();
            stack2.pop();
            return ret;
        } else {
            while(stack1.size() > 0){
                stack2.push(stack1.top());
                stack1.pop();
            }
            return pop();
        }
    }

    int top() {
        // write your code here
        if (stack2.size() > 0)
            return stack2.top();
        else
             while(stack1.size() > 0){
                stack2.push(stack1.top());
                stack1.pop();
            }
        return stack2.top();
    }
};
